File: leetcode/python/947.most-stones-removed-with-same-row-or-column.py
class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        n = len(stones)
        graph = [[] for _ in range(n)]

        for i in range(n):
            for j in range(i+1, n):
                if stones[i][0] == stones[j][0] or stones[i][1] == stones[j][1]:
                    graph[i].append(j)
                    graph[j].append(i)
        
        visited = set()

        def dfs(u: int):
            nonlocal visited
            for v in graph[u]:
                if v not in visited:
                    visited.add(v)
                    dfs(v)

        ans = n

        for u in range(n):
            if u not in visited:
                dfs(u)
                ans -= 1

        return ans

File: leetcode/rust-leetcode/src/947.most-stones-removed-with-same-row-or-column.rs
use std::collections::HashSet;
impl Solution {
    pub fn remove_stones(stones: Vec<Vec<i32>>) -> i32 {
        let n = stones.len();
        let mut set = HashSet::new();
        let mut graph = vec![vec![]; n];

        for i in 0..n {
            for j in i+1..n {
                if stones[i][0] == stones[j][0] || stones[i][1] == stones[j][1] {
                    graph[i].push(j);
                    graph[j].push(i);
                }
            }
        }

        fn dfs(graph: &Vec<Vec<usize>>, i: usize, set: &mut HashSet<usize>) {
            for &j in graph[i].iter() {
                if set.insert(j) {
                    dfs(graph, j, set);
                }
            }
        }

        let mut ans = n;
        for i in 0..n {
            if set.insert(i) {
                dfs(&graph, i, &mut set);
                ans -= 1;
            }
        }
        ans as _
    }
}

